src/python/pants/backend/python/rules/pytest_coverage.py

```diff
@@ -0,0 +1,415 @@
+#
+# Step 2.
+# Merging the Results: The `merge_coverage_data` rule executes `coverage combine`.
+#
+# Once we've run the tests, we have a bunch `TestResult`s, each with its own coverage data file, named `.coverage`.
+# In `merge_coverage_data` We stuff all these `.coverage` files into the same pex, which requires prefixing the
+# filenames with a unique identifier so they don't clobber each other. We then run
+# `coverage combine foo/.coverage bar/.coverage baz/.coverage` to combine all that data into a single .coverage file.
+#
+# Step 3.
+# Generating the Report: The `generate_coverage_report` rule executes `coverage html` or `coverage xml`
+#
+# Now we have one single `.coverage` file containing all our merged coverage data, with all the files referenced
+# as `foo/bar.py` and we want to generate a single report where all those files will be referenced with their
+# buildroot relative paths (`src/python/foo/bar.py`). Coverage requires that the files it's reporting on be
+# present in its environment when it generates the report, so we build a pex with all the source files with
+# their source roots intact, and *finally* our custom plugin starts doing some work. In our config file we've
+# given a map from sourceroot stripped file path to source root - `{'foo/bar.py': 'src/python`}` - As the report
+# is generated, every time coverage needs a file, it asks our plugin and the plugin says,
+#   "Oh, you want `foo/bar.py`? Sure! Here's `src/python/foo/bar.py`"
+# And then we get a nice report that references all the files with their buildroot relative names.
+#
+# Step 4.
+# Materializing the Report on Disk: The `run_tests` rule exposes the report to the user.
+#
+# Now we have a directory full of html or an xml file full of coverage data and we want to expose it to the user.
+# This step happens in `test.py` and should handle all kinds of coverage reports, not just pytest coverage.
+# The test runner grabs all our individual test results and requests a CoverageReport, and once it has one, it
+# writes it down in `dist/coverage` (or wherever the user has configured it.)
+
+
+COVERAGE_PLUGIN_MODULE_NAME = "__coverage_coverage_plugin__"
+
+DEFAULT_COVERAGE_CONFIG = dedent(
+    """
+    [run]
+    branch = True
+    timid = False
+    relative_files = True
+    """
+)
+
+
+def get_coveragerc_input(coveragerc_content: str) -> InputFilesContent:
+    return InputFilesContent([FileContent(path=".coveragerc", content=coveragerc_content.encode())])
+
+
+def get_coverage_plugin_input() -> InputFilesContent:
+    return InputFilesContent(
+        FilesContent(
+            (
+                FileContent(
+                    path=f"{COVERAGE_PLUGIN_MODULE_NAME}.py",
+                    content=pkg_resources.resource_string(__name__, "coverage_plugin/plugin.py"),
+                ),
+            )
+        )
+    )
+
+
+def get_packages_to_cover(
+    *, target: PythonTestsAdaptor, source_root_stripped_file_paths: Tuple[str, ...],
+) -> Tuple[str, ...]:
```

Comment:
For parity with the `generate_args` functions from the linters, consider changing `source_root_stripped_file_paths: Tuple[str, ...]` to `specified_source_files: SourceFiles`:

https://github.com/pantsbuild/pants/blob/c8959d23c94edfc76cb312ff8bb6fe22bf973feb/src/python/pants/backend/python/lint/flake8/rules.py#L39-L45